### ISSUE TITLE
LHN does not update properly when calling ReconnectApp

### ISSUE THREAD

**Issue by iwiznia** [Expensify/App#33771]
Coming from https://github.com/Expensify/App/issues/33372

# Problem

Left hand nav is outdated (in focus mode?) when calling ReconnectApp, which should never happen since after calling ReconnectApp everything should be up to date.

You can see logs from the App [here](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%2283d364e33dea345b-DFW%22)+AND+timestamp:[2023-12-29T15:24:35.186Z+TO+2023-12-29T17:24:35.186Z]&index=logs_expensify-026034) and the call to ReconnectApp [here](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%2283d2cfdb6887464d-DFW%22)+AND+timestamp:[2023-12-29T13:42:50.826Z+TO+2023-12-29T15:42:50.826Z]&index=logs_expensify-026033) and see that it failed to use the updates and instead called auth's `OpenApp`

My very wild guess is that for focus mode, it is never clearing the existing data in onyx, so a chat that was showing (since it was unread) would not get explicitly removed when OpenApp does not send it back.
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ea2b6c10b081507e</li>
        <li>Upwork Job ID: 1742324658321010688</li>
        <li>Last Price Increase: 2024-01-02</li>
    </ul>
</details>

**Comment by MitchExpensify** [comment]
What steps should I take to reproduce this @iwiznia ?

**Comment by marcaaron** [comment]
Probably it will be hard to reproduce since you need to miss x number of updates in order to trigger this part:

>it failed to use the updates and instead called auth's OpenApp

It will be easier on dev where we can just comment out some part of the code (not sure where exactly) and see what happens.

**Comment by marcaaron** [comment]
>for focus mode, it is never clearing the existing data in onyx, so a chat that was showing (since it was unread) would not get explicitly removed when OpenApp does not send it back

Different way to say the same thing

_while in focus mode, chats end up with stale data because an update from reliable updates queue is never received. If we fallback to OpenApp in this case we will send back only the reports for the focus mode query. Not the reports which the client has locally and something about them changed._

What is the best way to fix this?

As discussed, we think something like a `Onyx.setCollection()` could be the way forward. That way the fallback response of "OpenApp" can be used to overwrite existing values.

We will need to make sure that still works if you have pending actions. I think it would be fine as stuff like `AddComment` _should_ be queued to send first then `ReconnectApp` later.

**Comment by aldo-expensify** [comment]
> As discussed, we think something like a Onyx.setCollection() could be the way forward. That way the fallback response of "OpenApp" can be used to overwrite existing values.

This makes sense to me. An alternative I can think of is to send the reportIDs from the client, but I like that less and I guess it scales worse.

**Comment by marcaaron** [comment]
Ok, on a first pass - I cannot reproduce this problem and it does not seem like the current theory is the whole story.

If you look at the code for `ReconnectApp` we do not ever call it with the priority mode. 

**api.php**

https://github.com/Expensify/Web-Expensify/blob/bff668e4e707766249cddb1117b0a4caf89e1621/api.php#L3280

**AppInit.php (defaults to false)**

https://github.com/Expensify/Web-Expensify/blob/7938771c948ff4a3c7e16a101cdbe0a9e00dbf70/lib/AppInit.php#L368

Which means we should be syncing with the full list when GetOnyxUpdates throws the 423 jsonCode.

And when I test locally I see everything sync up fine 🤔

**Comment by aldo-expensify** [comment]
To reproduce and have `GetOnyxUpdates throws the 423 jsonCode`, do you disable reliable updates for the account?

**Comment by marcaaron** [comment]
No I just did this:

```diff
diff --git a/lib/AppInit.php b/lib/AppInit.php
index 52612dbdca9..049c39521f6 100644
--- a/lib/AppInit.php
+++ b/lib/AppInit.php
@@ -375,6 +375,7 @@ class AppInit
             /** @psalm-suppress InvalidGlobal */
             global $trackMetric;
             try {
+                throw new ExpError('asdf', 'asdf', jsonCode: 423);
                 $response = Auth::getOnyxUpdates($authToken, $updateIDFrom, $updateIDTo);

                 // GetOnyxUpdates will return the format array{accountID => array{updates, previousUpdateID, lastUpdateID}}.
```

**Comment by marcaaron** [comment]
This leaves me to think this _possibly_ has to do with the `mostRecentReportActionLastModified` changes which are still in effect when we call `ReconnectApp`.

**Reminder:** This pre-dates the "reliable updates" stuff where we will send only the new reports with based on the report actions that were last modified. That is very much based on the assumption that you would have everything that comes before this most recent update.

Also, note this logic does not change based on whether the priority mode filter is enabled or not... so again, I don't think this has to do with focus mode. Also, when you switch to Most recent mode we do a hard sync of the reports list (and do not pass the `mostRecentReportActionLastModified` in this case).

Maybe... somehow... a more recent report action sneaks in before the missing updates are fetched?

- User is offline
- User comes back online
- They somehow got an update - could it maybe be from a notification or something?
- State ends up with a gap between the new update and the fresh

Not sure exactly how it can happen, but that's my best improved theory.

**Comment by marcaaron** [comment]
Few more logs (not sure if they help):

**2023-12-29 14:42:32 707**

Message created by Garret here: https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%2283d2cf69cc627200-LHR%22)+AND+timestamp:[2023-12-29T13:42:32.708Z+TO+2023-12-29T15:42:32.708Z]&index=logs_expensify-026033

**2023-12-29 14:42:50 826**

@JmillsExpensify calls ReconnectApp [here](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%2283d2cfdb6887464d-DFW%22)+AND+timestamp:%5B2023-12-29T13:42:50.824Z+TO+2023-12-29T15:42:50.824Z%5D&index=logs_expensify-026033) unsuccessfully:

```
Processing 'ReconnectApp' for 'expensify.com' from '189.172.241.217' ~~ browserGUID: '653308122cd44' initialReferer: '<REDACTED>' command: 'ReconnectApp' policyIDList: '01DC7E46150451C4,03841C763918B9E1,063B20AF6F45E248,0BFB979BCD2DF930,0CFE78B25EBE2A0A,0D00F2B7A62842CD,137DA25D273F2423,2D3A0166B09CA932,39181563B22B74F3,3A89327178065942,5C853229E59E0C08,67B16182715EBC94,6B9A36D18CCA4933,72BEC40917763029,7A35FD885AE104E1,7DA2113291EE38AE,8C06041CC671A0EC,8C42263B74422F6A,9106A4CE5D9D5ACE,9132779747441AFB,94648D5F053554AD,9A3EEB48764CC921,A55A7962A1F001A4,A5F944A401046336,B648A8FBD506CAF9,B99FD813110CBCD3,BEDB558B1367FA21,BF5C59CAEE213690,C116E844BA7F5A40,E57898D9DE780B49' mostRecentReportActionLastModified: '2023-12-29 14:42:33.620' updateIDFrom: '66694010' appversion: '1.4.19-0' pusherSocketID: '633336.90971' authToken: '<REDACTED>' referer: 'ecash' platform: 'ios' api_setCookie: 'false' email: 'jason@expensify.com' partnerName: 'expensify.com' HTTP_CF_BOT_SCORE: '97' HTTP_CF_JA3_HASH: '773906b0efdefa24a7f2b8eb6985bf37'
```

The important part being:

```
mostRecentReportActionLastModified: '2023-12-29 14:42:33.620'
```

That's the same as the message from Garrett. The time of that log line was `2023-12-29 14:42:50 826` - so we can only get the updates from the past 17 seconds. Not sure what else to figure from this apart from that it confirms that Jason already had the message from Garrett in Onyx somehow. And as far as I can tell he got this update applied from another call to `ReconnectApp`.

This is the last successful time Jason got Onyx updates:

https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%2283cc3e066ec24754-DFW%22)+AND+timestamp:[2023-12-28T18:34:43.022Z+TO+2023-12-28T20:34:43.022Z]&index=logs_expensify-026023

`2023-12-28 19:34:43 019`

All times in between either his auth token was expired or the 423 was triggered. The trail is pretty hard to follow.

All I can advise for now is that we should either:

1. Stop using this `mostRecentReportActionLastModified` param. It is based on an assumption that could be false (even if we are not sure how it happens)
2. Figure out _how_ people are getting the partial updates and sending a bad `mostRecentReportActionLastModified` param.

**Comment by sakluger** [comment]
I don't think this is limited to focus mode.

1. I noticed that not all messages were showing up in my LHN, so I switched to most recent mode on NewDot chrome
2. I got a push notification on my iPhone for a new message in the chat (https://staging.new.expensify.com/r/1076415708177347) that I also had open on web (Chrome)
3. The message is not showing up in web

## Screenshots

Here is the push notification on mobile:

![Push notification](https://github.com/Expensify/App/assets/5997400/c77f08f6-c982-41e8-8a6f-55745d815eca)


Here's what I saw in web after receiving the push notification on mobile:

![image](https://github.com/Expensify/App/assets/5997400/9da35120-1ae2-402a-b88d-6f296e96b48a)

And here's what I saw in web after clicking to another chat then clicking back:

![image](https://github.com/Expensify/App/assets/5997400/e476862d-8bb6-495c-94bc-9856e883b33b)

**Comment by marcaaron** [comment]
@sakluger looks a bit different than this issue

Can't say with certainty but it seems like maybe:

- Message was sent fine via notification to your mobile
- Message did not make it to web (maybe it's slow for some reason or you lost connectivity for a moment)
- Tapped another chat and reliable updates sent you the missing update

**Comment by marcaaron** [comment]
Bringing the focus back to the main theory we have for this issue…

**Theory:**

`mostRecentReportActionLastModified` can somehow end up “newer” than the last modified for the “lastUpdateID” you have. When the 423 is handled we will only get everything from the `mostRecentReportActionLastModified` and there can be older updates you need.

Is there some way to test it? I can't really think of any way. Took a quick look at the `reliableUpdates` table and it has no `created` so we can't compare these two values easily.

**Comment by MitchExpensify** [comment]
Not overdue, Melvin

**Comment by iwiznia** [comment]
It kind of makes sense, but I am not that familiar with mostRecentReportActionLastModified and this flow

**Comment by marcaaron** [comment]
I don't know if there really is a way to repro, but if you look at the logs we can observe this:

- Clearly the `ReconnectApp` sync doesn't work in some cases (unclear why so far)
- It appears related to the `mostRecentReportActionLastModified`
- If you miss too many updates then you can only get whatever happened after the most recent report action update
- That's bad because the app does not actually "sync up" at all

Just my 2 cents. But now that we have the focus mode query AND reliable updates we might want to just give up on the `mostRecentReportActionLastModified` optimization. And should do something like this in the case where you can't get any missed updates:

- Assume the state of the app is unreliable and we should start from scratch
 - If you are on Focus mode
    - Dump all the reports (except the active one)
    - Have ReconnectApp get the Focus mode query reports again
- If you are on Most recent
    - Do the big OpenApp sync (same as when you log in)

**Comment by MitchExpensify** [comment]
Bit of a stretch but I wonder if this is related to this: https://github.com/Expensify/App/issues/34273 ("Message didn’t show up as unread in LHN, but I got the email notification")

**Comment by marcaaron** [comment]
Impossible to say without looking at a specific incident and scanning through the logs (which is very time consuming)

**Comment by marcaaron** [comment]
FWIW I think we should try to pop this in #vip-vsb or something. There is clearly some sort of hole that needs to be patched and a ton of context about how it might be happening. We should patch it so that the unexpected stuff we have observed can't happen.

**Comment by marcaaron** [comment]
Adding some more logs to help diagnose this. Also proposed to temporarily stop passing the `mostRecentReportActionLastModified` param [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1705610943122389?thread_ts=1705599372.427019&cid=C01GTK53T8Q) to potentially fix another issue. Not sure if it will solve this as I haven't fully reproduced the issue being experienced here - but have some hope that it will.

**Comment by sonialiap** [comment]
@marcaaron is LHN not updating report to read after report is opened by clicking a notification part of this issue? https://github.com/Expensify/App/issues/34698

**Comment by MitchExpensify** [comment]
@mallenexpensify Do you think this fits in wave 5 like https://github.com/Expensify/App/issues/34273 does?

**Comment by marcaaron** [comment]
@sonialiap possibly - but doesn't seem related to me. That one seems to imply something went wrong with marking chats as read.

**Comment by mallenexpensify** [comment]
The PR hit production 2 days ago, is anyone witnessing any changes/improvements? I don't think I have for the issue I'm tracking for 'LHN not updating for chats in #social'.  

@MitchExpensify I'm unsure if this is Wave 5, and/or if there's more work left to do.

**Comment by marcaaron** [comment]
`ReconnectApp` should be doing a full load now instead of passing the `mostRecentReportActionID` so I would expect the issue I explained [here](https://github.com/Expensify/App/issues/33771#issuecomment-1888133093) to now be 100% impossible. If there is still an issue then it's something else.

**Comment by mallenexpensify** [comment]
For #social, LHN still isn't updating with new chats, been testing all day.  Scott and Flo both reported not getting new chats too (which I consider the same and #social not uploading cuz the chat isn't updating in LHN, if it's a DM or a room)

**Comment by MitchExpensify** [comment]
> FWIW I think we should try to pop this in #vip-vsb

@quinthar, do you agree here?

**Comment by MitchExpensify** [comment]
Going to close this in the hope that https://github.com/Expensify/App/issues/33372 fixed it. Please feel free to reopen if anyone is experiencing it
